File: datnFE/src/models/banner.ts
import type { CommonStatus } from "./base";

export type BannerPosition =
  | "HOME_HERO"
  | "HOME_TOP"
  | "HOME_MIDDLE"
  | "HOME_BOTTOM"
  | "SIDEBAR"
  | "POPUP";

export type BannerType = "HERO" | "IMAGE" | "SLIDE";

export type LinkTarget = "NEW_TAB" | "SAME_TAB";

export interface BannerResponse {
  id: string;
  code: string;
  status: CommonStatus;
  title: string;
  subtitle?: string;
  description?: string;
  imageUrl: string;
  mobileImageUrl?: string;
  linkUrl?: string;
  linkTarget?: LinkTarget;
  position: BannerPosition;
  type: BannerType;
  priority?: number;
  startAt?: string;
  endAt?: string;
  buttonText?: string;
  backgroundColor?: string;
  createdDate: number;
  lastModifiedDate: number;
}

export interface BannerRequest {
  title: string;
  subtitle?: string;
  description?: string;
  imageUrl: string;
  mobileImageUrl?: string;
  linkUrl?: string;
  linkTarget?: LinkTarget;
  position: BannerPosition;
  type: BannerType;
  priority?: number;
  startAt?: string;
  endAt?: string;
  buttonText?: string;
  backgroundColor?: string;
}

export interface BannerSearchParams {
  page?: number;
  size?: number;
  keyword?: string;
  status?: CommonStatus;
  position?: BannerPosition;
  type?: BannerType;
}

export const initialBannerRequest: BannerRequest = {
  title: "",
  subtitle: "",
  description: "",
  imageUrl: "",
  mobileImageUrl: "",
  linkUrl: "",
  linkTarget: "SAME_TAB",
  position: "HOME_HERO",
  type: "IMAGE",
  priority: 0,
  startAt: undefined,
  endAt: undefined,
  buttonText: "",
  backgroundColor: "",
};

export const BANNER_POSITIONS: { value: BannerPosition; label: string }[] = [
  { value: "HOME_HERO", label: "Banner chính trang chủ" },
  { value: "HOME_TOP", label: "Banner phía trên trang chủ" },
  { value: "HOME_MIDDLE", label: "Banner giữa trang chủ" },
  { value: "HOME_BOTTOM", label: "Banner phía dưới trang chủ" },
  { value: "SIDEBAR", label: "Banner sidebar" },
  { value: "POPUP", label: "Banner popup" },
];

export const BANNER_TYPES: { value: BannerType; label: string }[] = [
  { value: "IMAGE", label: "Banner hình ảnh đơn" },
  { value: "HERO", label: "Banner dạng hero" },
  { value: "SLIDE", label: "Banner dạng slide/carousel" },
];

export const LINK_TARGETS: { value: LinkTarget; label: string }[] = [
  { value: "SAME_TAB", label: "Mở trong tab hiện tại" },
  { value: "NEW_TAB", label: "Mở trong tab mới" },
];

File: datnBE/src/main/java/com/example/datn/entity/Banner.java
package com.example.datn.entity;

import com.example.datn.entity.base.PrimaryEntity;
import com.example.datn.infrastructure.constant.BannerPosition;
import com.example.datn.infrastructure.constant.BannerType;
import com.example.datn.infrastructure.constant.EntityProperties;
import com.example.datn.infrastructure.constant.LinkTarget;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Table(name = "banner")
public class Banner extends PrimaryEntity {

    @Column(name = "title", length = EntityProperties.LENGTH_NAME, nullable = false)
    private String title;

    @Column(name = "subtitle", length = 255)
    private String subtitle;

    @Column(name = "description", length = 500)
    private String description;

    @Column(name = "image_url", length = EntityProperties.LENGTH_PICTURE, nullable = false)
    private String imageUrl;

    @Column(name = "mobile_image_url", length = EntityProperties.LENGTH_PICTURE)
    private String mobileImageUrl;

    @Column(name = "link_url", length = EntityProperties.LENGTH_URL)
    private String linkUrl;

    @Enumerated(EnumType.STRING)
    @Column(name = "link_target", length = 20)
    private LinkTarget linkTarget = LinkTarget.SAME_TAB;

    @Enumerated(EnumType.STRING)
    @Column(name = "position", length = 30, nullable = false)
    private BannerPosition position;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", length = 20)
    private BannerType type = BannerType.IMAGE;

    @Column(name = "priority")
    private Integer priority = 0;

    @Column(name = "start_at")
    private LocalDateTime startAt;

    @Column(name = "end_at")
    private LocalDateTime endAt;

    @Column(name = "button_text", length = 100)
    private String buttonText;

    @Column(name = "background_color", length = 20)
    private String backgroundColor;
}
